fwprovider/access/resource_realm.go

```diff
@@ -0,0 +1,431 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+package access
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/hashicorp/terraform-plugin-framework-validators/int32validator"
+	"github.com/hashicorp/terraform-plugin-framework-validators/resourcevalidator"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+
+	"github.com/bpg/terraform-provider-proxmox/fwprovider/config"
+	"github.com/bpg/terraform-provider-proxmox/proxmox"
+)
+
+var (
+	_ resource.Resource                     = (*realmResource)(nil)
+	_ resource.ResourceWithConfigure        = (*realmResource)(nil)
+	_ resource.ResourceWithImportState      = (*realmResource)(nil)
+	_ resource.ResourceWithConfigValidators = (*realmResource)(nil)
+)
+
+type realmResource struct {
+	client proxmox.Client
+}
+
+// NewRealmResource creates a new Realm resource.
+func NewRealmResource() resource.Resource {
+	return &realmResource{}
+}
+
+func (r *realmResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	cfg, ok := req.ProviderData.(config.Resource)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected config.Resource, got: %T", req.ProviderData),
+		)
+
+		return
+	}
+
+	r.client = cfg.Client
+}
+
+func (r *realmResource) ConfigValidators(_ context.Context) []resource.ConfigValidator {
+	return []resource.ConfigValidator{
+		resourcevalidator.Conflicting(
+			path.MatchRoot("realm"),
+		),
+	}
+}
+
+func (r *realmResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "Manages Realms on the Proxmox cluster",
+		MarkdownDescription: "Manages Realms on the Proxmox cluster.\n\n" +
+			"Realms are used for user authentication to a Proxmox node or cluster.\n",
+		Attributes: map[string]schema.Attribute{
+			"realm": schema.StringAttribute{
+				Description: "The ID of the realm which is of type string",
+				Required:    true,
+				Optional:    false,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"type": schema.StringAttribute{
+				Description: "Realm type",
+				Required:    true,
+				Optional:    false,
+				Validators: []validator.String{
+					stringvalidator.RegexMatches(
+						regexp.MustCompile(`^ad$|^ldap$|^openid$|^pam$|^pve$`),
+						"must be one of [`ad`, `ldap`, `openid`, `pam`, `pve`]",
+					),
```

Comment:
This would be more straightforward for "one of many" validation:
```suggestion
					stringvalidator.OneOf(`ad`, `ldap`, `openid`, `pam`, `pve`),
```

Reply:
same for `prompt`, `mode`, `sslversion`, etc attrs.

Reply:
Thanks for the hints!! These are areas of me using the generic approach to solving something because I lack the TF/golang specific knowledge at this point. Incredibly helpful pointers